docs/source/random_greedy_path.rst

```diff
@@ -0,0 +1,123 @@
+respect to the contraction with the 'best' cost. As such, contractions with
+very similar costs will be explored with equal probability, whereas those with
+higher costs will be less likely, but still possible. In this way, the
+optimizer can randomly explore the huge space of possible paths, but in a
+guided manner.
+
+The following graph roughly demonstrates the potential benefits of the
+``'random-greedy'`` algorithm, here for large randomly generated contractions,
+with either 8, 32 (the default), or 128 repeats:
+
+.. image:: path_found_flops_random.png
+    :width: 750
+
+.. note::
+
+    Bear in mind that such speed-ups are not guaranteed - it very much depends
+    on how structured or complex your contractions are.
+
+
+Customizing the Random-Greedy Path
+----------------------------------
+
+The random-greedy optimizer can be customized by instantiating your own
+:class:`~opt_einsum.path_random.RandomGreedy` object. Here you can control:
+
+* ``temperature`` - how far to stray from the locally 'best' contractions
+* ``rel_temperature`` - whether to normalize the temperature
+* ``nbranch`` - how many contractions (branches) to consider at each step
+* ``cost_fn`` - how to cost potential contractions
+
+There are also the main :class:`~opt_einsum.path_random.RandomOptimizer`
+options:
+
+* ``max_repeats`` - the maximum number of repeats
+* ``max_time`` - the maximum amount of time to run for (in seconds)
+* ``minimize`` - whether to minimize for total ``'flops'`` or ``'size'`` of the
+  largest intermediate
+
+For example, here we'll create an optimizer, then change its temperature
+whilst reusing it. We'll also set a high ``max_repeats`` and instead use a
+maximum time to terminate the search:
+
+.. code:: python
+
+    optimizer = oe.RandomGreedy(max_time=2, max_repeats=1_000_000)
+
+    for T in [1000, 100, 10, 1, 0.1]:
+        optimizer.temperature = T
+        path_rand_greedy = oe.contract_path(eq, *arrays, optimize=optimizer)[1]
+
+    print(len(optimizer.costs))  # the total number of trials so far
+    # 2555
+
+    print(math.log2(optimizer.best['flops']))
+    # 31.253582357751892
+
+So we have improved a bit on the standard ``'random-greedy'`` (which does 32
+repeats by default). The ``optimizer`` object now stores both the best path
+found so far - ``optimizer.path`` - as well as the list of flop-costs and
+maximum sizes found for each trial - ``optimizer.costs`` and
+``optimizer.sizes`` respectively.
+
+
+Parallelizing the Random-Greedy Search
+--------------------------------------
+
+Since each greedy attempt is independent, the random-greedy approach is
+naturally suited to parallelization. This is achieved by supplying a
+pool-executor like object, which should have an api matching the python 3
+`concurrent.futures <https://docs.python.org/3/library/concurrent.futures.html>`_
+module:
+
+.. code:: python
+
+    from concurrent.futures import ProcessPoolExecutor
+
+    executor = ProcessPoolExecutor()
```

Comment:
Here is there any reason not to use ThreadPoolExecutor? These should be self contained functions and the overhead is cheaper for thread pools.

Reply:
The problem is the GIL. Since greedy for example is essentially pure python, there is no meaningful parallelism unless we use processes. On the other hand, I have a prototype optimizer that uses [METIS](https://metis.readthedocs.io/en/latest/), and there, assuming most time is spent in the c-code, it would make sense to use a thread pool.

Reply:
Sorry, I always forget what the GIL does and does not lock as we usually thread in other languages. If the inputs are independent and/or copied, shouldn't the GIL release?

Reply:
Sadly no, GIL is by default always on even if all the data structures should be thread-safe. Its only really calls into other languages and IO etc. that threads offer speed-up for.